### DISCUSSION TITLE
How do I get dom element?

### DISCUSSION THREAD

**Discussion by MarcSkovMadsen** [jupyterlite/jupyterlite#112] [Q&A] [answered]
According to https://hacks.mozilla.org/2019/04/pyodide-bringing-the-scientific-python-stack-to-the-browser/ I should be able to `from js import document`. But I am not.

![image](https://user-images.githubusercontent.com/42288570/120080277-6111bd00-c0b8-11eb-827d-d836ec595026.png)

Related to https://github.com/jtpio/jupyterlite/discussions/112

**Comment by bollwyvl** [comment]
Due to its weight, the pyodide kernel runs inside a webworker, and doesn't
have direct access to the DOM. Some approaches we could try:

- offering a (default) worker kernel and a main thread kernel, with some
warnings
- jury-rigging a DOM proxy (something like this may already exist)

Either would be a fair bit of work, and neither will meet everyone's needs,
unfortunately.

**Reply by MarcSkovMadsen** [reply]
Thanks for taking the time.

If I could just get a dom elemnt selected and work with it, it would be so easy to provide all kindsof interactivity.

And there is no way to two-way communicate to js side either?

**Reply by MarcSkovMadsen** [reply]
Marc Skov Madsen 23:31
You can open a web socket though. So i guess the process is to 1) open s web socket 2) inject som javascript in a cell that will open connection to web socket. 3) use web socket for communication. I guess thats sort of rebuilding Bokeh :-)

**Reply by bollwyvl** [reply]
Yeah, the webworker boundary is pretty firmly enforced by the browser, and would need to be dissolved either by blocking the main thread every time the WASM stuff takes over, or emulating the calls with a [proxy](https://github.com/mmis1000/DOM-Proxy). 

As you noted on the linked issue, the comms approach (#18) is the most likely to be implemented, as it will work with the most existing things, though there will be a _lot_ of stack work required to trick various packages into installing from `micropip` without e.g. zmq... we'll probably need some shim packages, but again, it should be a countable number but would unlock a great deal of existing technology. For example, the bokeh->dask->distributed stack is really just blocked by a pyodide-WASM-native build of `libyaml`. 

In the interim, you do have _some_ of the regular "abuse" tools in your arsenal, though there is no support yet for `update_display`, etc. But at least you'd be able to do 

```py
 In[1]:  from pyolite.display import * # vs from IPython.display import *

 In[2]:  display(HTML('<h1 id="woo">WOO</h1>'))

Out[2]: WOO

 In[3]:  display(HTML('<script>document.querySelector("#woo").textContent = "WOOHOO"</script>'))

# Out[2] will change to WOOHOO
```

There _is_ two-way communication, but nobody is _listening_ on the other side, really:

```py
import js
js.self.postMessage('woohoo')
```

**Reply by jtpio** [reply]
Maybe the same `pyolite` kernel could also be made available (as 2 separate kernels showing up in the UI):

- running in a web worker (like it is now)
- running in the main thread

The startup logic would be different, but the main execution part would be the same.

Also for reference Basthon JupyterLab starts the Basthon kernel in the main thread with access to the DOM: https://framagit.org/casatir/basthon-jupyterlab/

For now it's a fork of JupyterLite, but when the `jupyterlite` tooling is in place this kernel could be implemented and shipped as a federated extension.

**Comment by MarcSkovMadsen** [comment]
Thanks. So my question now is. Could I use `display(HTML(...` to setup communication? And how? And would that be a good approch?  Or would it be better to use web socket?

And would it be possible to use the `display(HTML(...` to grab the WebWorker and configure the `onmessage` callback?

**Comment by MarcSkovMadsen** [comment]
I added a Feature request in Bokeh to add support for Jupyterlite. 

https://github.com/bokeh/bokeh/issues/11315

It was closed down with the comment that a Bokeh wheel is coming in Bokeh 2.4. As far as I understand that would probably not be enough.

If you would add any knowledge you have like "For example, the bokeh->dask->distributed stack is really just blocked by a pyodide-WASM-native build of libyaml." in the Bokeh discussion, it would be very much appreciated. 

There are so many things depending on Bokeh that it is key to get it working on @pyodide and Jupyterlite as I see it.

Thanks.

**Comment by jtpio** [comment, marked as answer]
There is also some initial work in pyodide to add support for `Comlink` to improve the web worker scenario: https://github.com/pyodide/pyodide/pull/1535 (also related: https://github.com/pyodide/pyodide/issues/1518).
